rfcs/20210510-binary-size-reduction/ISA_extension.md

```diff
@@ -0,0 +1,119 @@
+# Build knobs controlling binary size extension.
+
+## Problem Statement.
+This follows up this
+[RFC](https://github.com/oneapi-src/oneDNN/tree/rfcs/rfcs/20210510-binary-size-reduction)
+and proposes additional build knob to control functionality by ISA. It may
+provide additional save up to 5 MB, which is around ten percent (~10%) from
+current full binary size, based on selected ISA.
+
+## Proposal
+The proposal is to introduce one more build option,
+`DNNL_ENABLE_PRIMITIVE_CPU_ISA`. It would take effect if `DNNL_X64` build
+setting was specified.
+
+Library supported ISA are only partially ordered, that is why the suggestion is
+to group them, so that these groups are in linear order. Here they are:
+* `SSE41`  - supports sse41 ISA only and all compiler-based optimized
+             implementations.
+* `AVX2`   - supports `SSE41` group and avx and avx2 ISA extensions on top.
+* `AVX512` - supports `AVX2` group and most AVX512 ISA extensions: avx512_mic,
+             avx512_mic_4ops, avx512_core, avx512_core_vnni, avx512_core_bf16
+             and avx2_vnni.
+* `AMX`    - supports `AVX512` group and avx512_core_amx ISA extension on top.
+* `ALL`    - is an alias to the latest group (supports all groups altogether).
```

Comment:
Hi @dzarukin 
I was wondering that if one wants to target Alderlake (`AVX2_VNNI`) then effectively we have only `ALL` as the valid option. Will it be better to include all the ISA (as they are supported in CPU dispatch control) options?

Reply:
Or maybe `AVX2` option can also include `avx_vnni`

Reply:
@alokbakshi, thank for the question. We considered VNNI as AVX512 extension (see line 22). Despite the fact it is attached to AVX2, this is a modern hardware and this change targets to save up for really old ISA like pure AVX2. Thanks.